### ISSUE TITLE
[Bug]: Re-tiling issues

### ISSUE THREAD

**Issue by tvarsis** [zeroxoneafour/polonium#157]
### Before you make the issue, confirm that

- [X] You are running the master build from GitHub actions
- [X] You have read the FAQ and the usage instructions
- [X] You have searched pull requests and issues and have found none similar to your issue
- [X] This problem comes from a failure in the tiling system and not a missing feature

### Describe the bug

The re-tiling shortcut woks sometimes, but most of the time it does nothing. Same happens when you connect/disconnect an external monitor with different resolution, leading to broken tiling which is hard to re-tile, even manually. Usually you would have to close and reopen windows to get it to tile correctly. 

### To reproduce

There are multiple ways to reproduce this one, but it would happen after a tiling state becomes broken and you would expect the "re-tile" shortcut to fix it, but it doesn't. One way is to have 3 windows tiled correctly and then connect an external monitor in the mode that you only have output on the external monitor and not the laptop screen. Tiling will become broken, and re-tile shortcut does nothing.

### Expected behavior

Pressing the re-tile shortcut should always force a re-tile of all windows on the active desktop to the selected tiled layout. Connecting or disconnecting an external monitor should also automatically force a re-tile of all windows on the active desktop to avoid broken states. It seems to be in very rare cases where the re-tile shortcut actually works as expected.

### Log output

```shell
Nothing in the logs
```

**Comment by zeroxoneafour** [comment]
https://zeroxoneafour.github.io/polonium/usage.html#getting-a-log

**Comment by root-hal9000** [comment]
I will add that this happens to me sometimes, but after pressing the retiling shortcut, if I use another shortcut to move the window, it will get back into a tile (shortcuts like the polonium one for insert into or the kwin shortcut for move one screen to the left/right)

**Comment by tvarsis** [comment]
@zeroxoneafour Added logs now. The logs are for a scenario where I have one Firefox window tiled to the left (half screen) and adding a new Firefox window that then opens in full screen mode getting placed over the other FF window. Then I try to press Retile shortcut, but nothing happens. Expected result is for it to auto tile directly on open, but if that does not work, then the retile should sort it out.  Not even trying to manually tile, drag or using "insert" shortcuts. Seems like it has a tendency to get into broken state where it is not possible to retile, but you would have to basically close and reopen all window. A "force retile all open windows" shortcut would be nice to have.

**Comment by root-hal9000** [comment]
hey @tvarsis , just curious, but does Firefox seem to behave worse than others? I can't quantify it, but I feel like I always have more issues with Firefox than other applications - particularly when you create a new window by dragging a tab out of an existing window

**Comment by tvarsis** [comment]
@root-hal9000 Yeah, firefox and electron/flatpaks seems to work worst. But I gave up and switched to Krohnkite that have been updated for KDE 6 and works much better and more expected.

**Comment by zeroxoneafour** [comment]
Sometimes this does happen to me. I find that the fastest solution is quickly flicking between desktops. Unfortunately, I have not been using Linux much as of recent due to a wide variety of crashes, kernel panics etc. I also really doubt that an easy solution really exists for this problem.

**Comment by zeroxoneafour** [comment]
Obviously the issue _is_ inherent to kwin. There's no code in Polonium that checks if an application is firefox and decides to not tile it. There's no code in Khronkite that checks if an application is firefox and decides to tile it. It just happens that Khronkite manages its own tiles and Polonium relies on KWin to handle a lot of the tiling stuff. Theoretically, everything should work. In reality, Khronkite has a lot more fine-grained control over application sizing and tiling and doesn't have to believe in KWin's ability to move a window into a tile when told.

Polonium is deliberately an experiment to implement tiling on and fully utilizing the KWin tiling API, which is iffy more times than it perhaps should be. If Khronkite works better for you, then use it.

**Comment by jay-depot** [comment]
> This problem comes from a failure in the tiling system and not a missing feature

I'm not 100% convinced this is properly a bug, rather than a missing feature. The default "re-tile" keyboard shortcut for Polonium is labeled in the settings window as affecting only the current window. I think it would be more accurate to say this is a request for a new feature, specifically a "re-tile all windows (on current desktop)" shortcut. The real question is whether that's possible using kwin's current tiling API.

**Comment by Raven-002** [comment]
>  I think it would be more accurate to say this is a request for a new feature, specifically a "re-tile all windows (on current virtual desktop)" shortcut

I've added a `restart engine` shortcut (#180) and some other improvements. testing the behavior described in this thread, of opening 3 windows and then connecting an external monitor, the bugs still occur to me, but the `restart engine` shortcut retile all windows without a problem (from the testing i did)

You can try the version with all of my changes now here: [https://github.com/Raven-002/polonium/tree/dev](https://github.com/Raven-002/polonium/tree/dev). Run `make install` and then relog or run `kwin_wayland --replace` (this will close all windows) or wait for the PRs to be merged

It is not the best solution, but it is a simple fix that solves almost all of the issues i had
